Engine/src/Input/Input.bf

```diff
@@ -10,40 +11,121 @@ namespace SteelEngine.Input
+				delete pad;
+		}
 
 		static void KeyEvent(KeyCode kc, KeyEvent ke)
 		{
 			_accumulatedEvents[kc.Underlying] |= ke;
 		}
 
+		static void AxisEvent(AxisCode ac, float value)
+		{
+			_accumulatedAxisValues[ac.Underlying] += value;
+		}
+
+		static void GamepadKeyEvent(GamepadId id, KeyCode kc, KeyEvent ke)
+		{
+			GamepadInfo gamepad;
+			if (_gamepads.TryGetValue(id, out gamepad))
+			{
+				gamepad.[Friend]SetKey(kc, ke);
+			}
+			KeyEvent(kc, ke);
+		}
+
+		static void GamepadAxisEvent(GamepadId id, GamepadAxisCode ac, float value)
+		{
+			GamepadInfo gamepad;
+			if (_gamepads.TryGetValue(id, out gamepad))
+			{
+				gamepad.[Friend]SetAxis(ac, value);
+			}
+			AxisEvent((.)ac, value);
+		}
+
+		static void UpdateMousePosition(float x, float y)
+		{
+			_mousePos = (x,y);
+		}
+
+		static void GamepadConnected(GamepadId gamepadId, StringView deviceName)
+		{
+			_gamepads.Add(gamepadId, new GamepadInfo(deviceName));
+		}
+
+		static void GamepadDisconnected(GamepadId gamepadId)
+		{
+			if (_gamepads.GetAndRemove(gamepadId) case .Ok(let val))
+			{
+				delete val.value;
+			}
+		}
+
 		static void Update()
 		{
+			// Update keys
 			for (int i = 0, let count = _accumulatedEvents.Count; i < count; i++)
 			{
 				let event = _accumulatedEvents[i];
+				var newValue = _lastUpdateState[i] & ~(.Down | .Up);	// Clear up / down flags
 
-				_lastUpdateState[i] &= ~(.Down | .Up);	// Clear up / down
-				if (event.HasFlag(.Down))
+				// If we already have key state Hold we don't want to set Down again
+				if (event.HasFlag(.Down) && !_lastUpdateState[i].HasFlag(.Hold))
 				{
-					_lastUpdateState[i] = .Down | .Hold;
+					newValue = .Down | .Hold;								
 				}
 
-				if (event.HasFlag(.Up))
+				// Up key flag can only be set when if the key was down previous update
+				if (event.HasFlag(.Up) && _lastUpdateState[i].HasFlag(.Hold))
 				{
-					_lastUpdateState[i] = ( _lastUpdateState[i] & ~.Hold) | .Up;		// If key was pressed and released we probably want to unset the Hold flag but keep down flag.
+					// If key was pressed and released we probably want to unset the Hold flag but keep down flag.
+					newValue = ( _lastUpdateState[i] & ~.Hold) | .Up;
 				}
+
+				_lastUpdateState[i] = newValue;
 			}
 
-			// Clear event accumulator
+			// Update gamepads
+			for (let pad in _gamepads.Values)
+				pad.Update();
+
+			// Update axis
+			_axisValues = _accumulatedAxisValues;
+
+			// Update mouse
+			_mouseDelta = (_mousePos.0 - _lastMousePos.0, _mousePos.1 - _lastMousePos.1);
+			_axisValues[(int)AxisCode.MouseX] = _mouseDelta.0;
+			_axisValues[(int)AxisCode.MouseY] = _mouseDelta.1;
+			_lastMousePos = _mousePos;
+
+			// Clear accumulators
 			_accumulatedEvents = default;
+			_accumulatedAxisValues = default;
 		}
 
 		public static void ResetInput()
 		{
 			_accumulatedEvents = default;
 			_lastUpdateState = default;
+			_axisValues = default;
+			for (let pad in _gamepads.Values) pad.ResetInput();
```

Comment:
To be consistent with other one-liner for loops, move `pad.ResetInput()` to the next line.